### DISCUSSION TITLE
Read Tool Enhancements

### DISCUSSION THREAD

**Discussion by CiberNin** [cline/cline#905] [Feature Requests]
**Feature Request: Read Tool Enhancements**

Extend the `<read_file>` tool to accept multiple `<path>` inputs simultaneously. This would reduce round trips and improve efficiency, particularly when reading entire projects or folders. It would also optimize token usage, which, while cached tokens are relatively cheap, still accumulate a cost under the current one-at-a-time system.

For safety, it may make sense to add a configurable limit in tokens, bytes, or lines of code after which the user will be prompted for approval. Which to be useful needs modified to list out all files that will be read, their individual size, and also the total size of the read (since that's what triggers human review in the first place.

**Further Work**
1. Support bare folder paths, where all files (and sub-files?) within the folder get read. 100% definitely need the safety setting for this one.
2. When asking human for read approval, maybe have a checkbox next to each filename so they can specifically filter out one or two files they don't want read? 
3. Could fallback to listing metadata for binary files.
4. Updating `<list_files>` tool & the environment data (they use same function) to list file / folder sizes could help Roo intelligently decide which files to read.
5. Could keep track of if/when each file was last read / if its still in context (has not been pruned). And then if a write is attempted on a file, auto reject with the current contents of the file if it's latest version is not in memory. I don't know about anyone else, but I fairly often have Roo try to write to a file that already exists which it did not bother to read. Or I update a file manually after Roo does & it does not know it needs to check my updates.

**Comment by BradKML** [comment]
What would be the major complication for this feature? RooCode also had the same ideas but not that many did this

**Comment by d0lwl0b** [comment]
are you interested in #5912 ?
